File: Marmalade/MarmaladeODK/source/android/CallbacksRequestGamerInfo.cpp
/*
 * Copyright (C) 2012, 2013 OUYA, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "CallbacksRequestGamerInfo.h"
#include "CallbackSingleton.h"

#include "ODK_internal.h"
#include "s3eEdk.h"
#include "s3eEdk_android.h"
#include "IwDebug.h"

#include <stdio.h>

void CallbacksRequestGamerInfo::RegisterCallback(s3eCallback callback, s3eCallback* savedCallback, int callbackType)
{
	if (*savedCallback)
	{
		//IwTrace(ODK, ("Unregistering Callback"));

		s3eEdkCallbacksUnRegister(
				S3E_EXT_ODK_HASH,
				S3E_ODK_CALLBACKS_MAX,
				callbackType,
				*savedCallback,
				NULL);
		*savedCallback = NULL;
	}

	*savedCallback = callback;

	//IwTrace(ODK, ("Registering Callback"));

	s3eEdkCallbacksRegister(
			S3E_EXT_ODK_HASH,
			S3E_ODK_CALLBACKS_MAX,
			callbackType,
			*savedCallback,
			NULL,
			S3E_FALSE);
}

void CallbacksRequestGamerInfo::RegisterCallbacks(s3eCallback onSuccess, s3eCallback onFailure, s3eCallback onCancel)
{
	RegisterCallback(onSuccess, &m_onSuccess, S3E_ODK_CALLBACKS_REQUEST_GAMER_INFO_ON_SUCCESS);
	RegisterCallback(onFailure, &m_onFailure, S3E_ODK_CALLBACKS_REQUEST_GAMER_INFO_ON_FAILURE);
	RegisterCallback(onCancel, &m_onCancel, S3E_ODK_CALLBACKS_REQUEST_GAMER_INFO_ON_CANCEL);
}

void CallbacksRequestGamerInfo::OnSuccess(const OuyaSDK::GamerInfo& gamerInfo)
{
	//char buffer[256];
	//sprintf(buffer, "OnSuccess: username=%s uuid=%s", gamerInfo.Username.c_str(), gamerInfo.Uuid.c_str());
	//IwTrace(ODK, (buffer));

	s3eRequestGamerInfoSuccessEvent event;
	event.m_gamerInfo.Init();
	event.m_gamerInfo.Copy(gamerInfo);

	m_dataRequestGamerInfoSuccessEvent = event; //don't send a temp pointer
	s3eEdkCallbacksEnqueue(S3E_EXT_ODK_HASH, S3E_ODK_CALLBACKS_REQUEST_GAMER_INFO_ON_SUCCESS, &m_dataRequestGamerInfoSuccessEvent);
}

void CallbacksRequestGamerInfo::OnFailure(int errorCode, const std::string& errorMessage)
{
	char buffer[256];
	sprintf(buffer, "OnFailure errorCode=%d errorMessage=%s", errorCode, errorMessage.c_str());
	IwTrace(ODK, (buffer));

	s3eRequestGamerInfoFailureEvent event;
	event.m_errorCode = errorCode;
	sprintf(event.m_errorMessage, "%s", errorMessage.c_str());

	m_dataRequestGamerInfoFailureEvent = event; //don't send a temp pointer
	s3eEdkCallbacksEnqueue(S3E_EXT_ODK_HASH, S3E_ODK_CALLBACKS_REQUEST_GAMER_INFO_ON_FAILURE, &m_dataRequestGamerInfoFailureEvent);
}

void CallbacksRequestGamerInfo::OnCancel()
{
	IwTrace(ODK, ("OnCancel"));

	s3eRequestGamerInfoCancelEvent event;

	m_dataRequestGamerInfoCancelEvent = event; //don't send a temp pointer
	s3eEdkCallbacksEnqueue(S3E_EXT_ODK_HASH, S3E_ODK_CALLBACKS_REQUEST_GAMER_INFO_ON_CANCEL, &m_dataRequestGamerInfoCancelEvent);
}

File: Marmalade/MarmaladeODK/source/android/CallbacksRequestGamerInfo.java
/*
 * Copyright (C) 2012, 2013 OUYA, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package tv.ouya.sdk.marmalade;


import android.util.Log;


public class CallbacksRequestGamerInfo {

	private final String LOG_TAG  = CallbacksRequestGamerInfo.class.getSimpleName();

	public native void CallbacksRequestGamerInfoOnSuccess(String jsonData);
	public native void CallbacksRequestGamerInfoOnFailure(int errorCode, String errorMessage);
	public native void CallbacksRequestGamerInfoOnCancel();

	public void onSuccess(final String jsonData) {
		Log.i(LOG_TAG, "onSuccess=" + jsonData);
		CallbacksRequestGamerInfoOnSuccess(jsonData);
	}

	public void onFailure(final int errorCode, final String errorMessage) {
		Log.i(LOG_TAG, "onFailure: errorCode=" + errorCode + " errorMessagee=" + errorMessage);
		CallbacksRequestGamerInfoOnFailure(errorCode, errorMessage);
	}

	public void onCancel() {
		Log.i(LOG_TAG, "onCancel");
		CallbacksRequestGamerInfoOnCancel();
	}
}